src/main/java/org/apache/sling/jcr/repoinit/impl/RepositoryInitializerFactory.java

```diff
@@ -122,14 +124,58 @@ public void processRepository(final SlingRepository repo) throws Exception {
                             continue;
                         }
                         final List<Operation> ops = parser.parse(new StringReader(script));
-                        log.info("Executing {} repoinit operations", ops.size());
-                        processor.apply(s, ops);
-                        s.save();
+                        String msg = String.format("Executing %s repoinit operations", ops.size());
+                        log.info(msg);
+                        applyOperations(s,ops,msg);
                     }
                 }
             } finally {
                 s.logout();
             }
         }
     }
+
+
+    /**
+     * Apply the operations within a session, support retries
+     * @param session the JCR session to use
+     * @param ops the list of operations
+     * @param logMessage the messages to print when retry
+     * @throws Exception if the application fails despite the retry
+     */
+    private void applyOperations(Session session, List<Operation> ops, String logMessage) throws RepositoryException {
+
+        RetryableOperation retry = new RetryableOperation.Builder().withBackoffBaseMsec(1000).withMaxRetries(3).build();
+        RetryableOperation.RetryableOperationResult result = retry.apply(() -> {
+            try {
+                processor.apply(session, ops);
+                session.save();
+                return new RetryableOperation.RetryableOperationResult(true,false,null);
+            } catch (InvalidItemStateException ise) {
+                // a retry makes sense, because this exception might be caused by an concurrent operation
+                log.debug("(temporarily) failed to apply repoinit operations",ise);
+                try {
+                    session.refresh(false); // discard all pending changes
+                } catch (RepositoryException e1) {
+                    // ignore
+                }
+                return new RetryableOperation.RetryableOperationResult(false,true,ise);
+            } catch (RepositoryException re) {
+                // a permanent error, retry is not useful
+                try {
+                    session.refresh(false); // discard all pending changes
+                } catch (RepositoryException e1) {
+                    // ignore
+                }
+                return new RetryableOperation.RetryableOperationResult(false,false,re);
+            }
+        }, logMessage);
+        if (!result.isSuccessful()) {
```

Comment:
If retries were needed, it would be good to log their count at the DEBUG level.

Reply:
This is done as part of the RetryableOperation[1].

[1] https://github.com/apache/sling-org-apache-sling-jcr-repoinit/blob/feature/SLING-10418-retry/src/main/java/org/apache/sling/jcr/repoinit/impl/RetryableOperation.java#L60